File: swarmmanagercontroller/src/main/java/com/swarmmanager/docker/api/common/client/jaxrs/filter/ResponseStatusExceptionFilter.java
package com.swarmmanager.docker.api.common.client.jaxrs.filter;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.swarmmanager.docker.api.common.client.jaxrs.exception.DockerRemoteApiException;
import org.apache.commons.io.IOUtils;

import javax.ws.rs.client.ClientRequestContext;
import javax.ws.rs.client.ClientResponseContext;
import javax.ws.rs.client.ClientResponseFilter;
import javax.ws.rs.core.MediaType;
import java.io.EOFException;
import java.io.IOException;
import java.io.InputStream;
import java.nio.charset.Charset;

public class ResponseStatusExceptionFilter implements ClientResponseFilter {

    @Override
    public void filter(ClientRequestContext requestContext, ClientResponseContext responseContext) throws IOException {
        int status = responseContext.getStatus();
        // status code used by the frontend to identify docker api errors
        int dockerErrStatus = 100001;
        if (status >= 300) {
            if (status == 503) {
                throw new DockerRemoteApiException("Configured node is not part of a swarm.", dockerErrStatus);
            }
            throw new DockerRemoteApiException(getBodyAsMessage(responseContext), dockerErrStatus);
        }
    }

    private String getBodyAsMessage(ClientResponseContext responseContext) throws IOException {
        if (responseContext.hasEntity()) {
            int contentLength = responseContext.getLength();
            if (contentLength != -1) {
                byte[] buffer = new byte[contentLength];
                try {
                    InputStream entityStream = responseContext.getEntityStream();
                    IOUtils.readFully(entityStream, buffer);
                    entityStream.close();
                } catch (EOFException e) {
                    return null;
                }
                Charset charset = null;
                MediaType mediaType = responseContext.getMediaType();
                if (mediaType != null) {
                    String charsetName = mediaType.getParameters().get("charset");
                    if (charsetName != null) {
                        try {
                            charset = Charset.forName(charsetName);
                        } catch (Exception e) {
                            // Do noting...
                        }
                    }
                }
                if (charset == null) {
                    charset = Charset.defaultCharset();
                }
                ObjectMapper mapper = new ObjectMapper();
                JsonNode node = mapper.readTree(new String(buffer, charset));
                String message = node.get("message").asText();
                if (message != null) {
                    String[] parts = message.split(" ");
                    if (parts.length > 0 && parts[0] != null && parts[0].length() > 0) {
                        parts[0] = parts[0].substring(0, 1).toUpperCase() + parts[0].substring(1);
                        message = String.join(" ", parts);
                    }
                }
                return message;
            }
        }
        return null;
    }
}

File: swarmmanagercontroller-new/src/main/kotlin/com/swarmbit/docker/api/common/client/jaxrs/filter/ResponseStatusExceptionFilter.kt
package com.swarmbit.docker.api.common.client.jaxrs.filter

import com.fasterxml.jackson.databind.ObjectMapper
import com.swarmbit.docker.api.common.exception.DockerRemoteApiException
import mu.KotlinLogging
import java.io.IOException
import javax.annotation.Priority
import javax.ws.rs.client.ClientRequestContext
import javax.ws.rs.client.ClientResponseContext
import javax.ws.rs.client.ClientResponseFilter

private val logger = KotlinLogging.logger {}
private val objectMapper = ObjectMapper()

@Priority(5000)
class ResponseStatusExceptionFilter : ClientResponseFilter {

    @Throws(IOException::class)
    override fun filter(requestContext: ClientRequestContext, responseContext: ClientResponseContext) {
        val status = responseContext.status
        if (status >= 300) {
            if (status == 503) {
                throw DockerRemoteApiException("Configured node is not part of a swarm.", status)
            }
            var errorString = "Docker api error"
            if (responseContext.hasEntity()) {
                val entityString = readEntity(responseContext)
                errorString = objectMapper.readTree(entityString)["message"].asText()
                logger.debug { "DOCKER API ERROR: $entityString" }
            }
            throw DockerRemoteApiException(errorString, status)
        }
    }

}